pkg/ovsdb/monitors.go

```diff
@@ -5,198 +5,262 @@ import (
-		handlers := m.modify[&m.updaters[indx]]
-		for k, v := range handlers {
-			if v == handler {
-				handlers[k] = handlers[len(handlers)-1]
-				m.modify[&m.updaters[indx]] = handlers[:len(handlers)-1]
+			if len(m.upater2handlers[updaterKey]) == 0 {
+				delete(m.upater2handlers, updaterKey)
+				updt := m.key2Updaters[table]
+				for i, v := range updt {
+					if v.key == updaterKey {
+						m.key2Updaters[table] = append(updt[:i], updt[i+1:]...)
+						break
+					}
+				}
 			}
 		}
 	}
-	if len(m.modify[&m.updaters[indx]]) == 0 &&
-		len(m.delete[&m.updaters[indx]]) == 0 &&
-		len(m.insert[&m.updaters[indx]]) == 0 {
-		// we can remove this handler
-		m.updaters[indx] = m.updaters[len(m.updaters)-1]
-		m.updaters = m.updaters[:len(m.updaters)-1]
-		if len(m.updaters) == 0 {
-			m.cancel()
-			return true
-		}
-	}
-	return false
+	return
 }
 
-func newMonitor(cli *clientv3.Client, prefix string, mcr ovsjson.MonitorCondRequest, isV1 bool, handler *handlerKey) *monitor {
-
-	ctx := context.Background()
-	ctxt, cancel := context.WithCancel(ctx)
-	wch := cli.Watch(clientv3.WithRequireLeader(ctxt), prefix,
-		clientv3.WithPrefix(),
-		clientv3.WithCreatedNotify(),
-		clientv3.WithPrevKV())
-
-	monitor := &monitor{cancel: cancel}
+func (m *monitor) start() {
 	go func() {
-		for wresp := range wch {
-			for _, ev := range wresp.Events {
-				var dest map[*updater][]*handlerKey
-				monitor.mu.Lock()
-				if ev.IsCreate() {
-					// insert event
-					dest = monitor.insert
-				} else if ev.IsModify() {
-					dest = monitor.modify
-				} else {
-					// delete event
-					dest = monitor.delete
+		for wresp := range m.watchChannel {
+			if wresp.Canceled {
+				// TODO should we just reconnect
+				m.mu.Lock()
+				for hlk := range m.handlers {
+					// run in separate goroutines
+					go hlk.handler.monitorCanceledNotification(hlk.jsonValue)
 				}
-				monitor.mu.Unlock()
-				// TODO queue or a separate goroutine
-				monitor.propagateEvent(ev, dest)
+				m.mu.Unlock()
+				// remove itself
+				m.db.RemoveMonitor(m.dataBaseName)
+				return
+			}
+			result, _ := m.prepareTableUpdate(wresp.Events)
+			for hd, tu := range result {
+				go hd.handler.notify(hd.jsonValue, tu)
 			}
 		}
 	}()
-	return monitor
 }
 
-func (m *monitor) propagateEvent(event *clientv3.Event, destinations map[*updater][]*handlerKey) {
-	// for each updater entry create a return data and call each handler
+func (m *monitor) hasHandlers() bool {
+	return len(m.handlers) > 0
 }
 
-func (u *updater) prepareTableUpdate(event *clientv3.Event) (table string, uuid string, rowUpdate *ovsjson.RowUpdate, err error) {
-	key := string(event.Kv.Key)
-	if table, uuid, err = keyToTableAndUUID(key); err != nil {
-		return
+func mcrToUpdater(mcr ovsjson.MonitorCondRequest, isV1 bool) *updater {
+	sort.Strings(mcr.Columns)
+	var key string
+	for _, c := range mcr.Columns {
+		key = key + c
+	}
+	// TODO handle "Where"
+	if mcr.Select == nil {
+		mcr.Select = &allTrueSelect
+	}
```

Comment:
I think it is per each option independantly, so whatever is missing get the value of true:

```
   o  If "initial" is omitted or true, every row in the table is sent as
      part of the response to the "monitor" request.

   o  If "insert" is omitted or true, "update" notifications are sent
      for rows newly inserted into the table.

   o  If "delete" is omitted or true, "update" notifications are sent
      for rows deleted from the table.

   o  If "modify" is omitted or true, "update" notifications are sent
      whenever a row in the table is modified.
```